### PR TITLE
feat(planner): Implement cost-based optimization

### LINKED ISSUE #6643: Feature: Cost-based optimization(CBO)
Support cost-based optimization.

Need RFC.

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Close #6643

### CODE REVIEW

**Comment by drmingdrmer** [comment]
- `main` branch is force pushed, update your working branch please: #7261

File: src/query/datavalues/src/data_value.rs [hunk 1]
```diff
@@ -367,6 +368,23 @@ impl Ord for DataValue {
     }
 }
 
+#[allow(clippy::derive_hash_xor_eq)]
+impl Hash for DataValue {
```
Comment (leiysky) [inline src/query/datavalues/src/data_value.rs:372]: @sundy-li  This PR implemented `Hash` for `DataValue` and `DataTypeImpl`, PTAL.

File: src/query/service/src/sql/planner/plans/limit.rs [hunk 2]
```diff
@@ -71,6 +71,15 @@ impl PhysicalOperator for Limit {
...
 impl LogicalOperator for Limit {
     fn derive_relational_prop<'a>(&self, rel_expr: &RelExpr<'a>) -> Result<RelationalProperty> {
-        rel_expr.derive_relational_prop_child(0)
+        let input_prop = rel_expr.derive_relational_prop_child(0)?;
+
+        Ok(RelationalProperty {
+            output_columns: input_prop.output_columns,
+            outer_columns: input_prop.outer_columns,
+            cardinality: match self.limit {
+                Some(limit) if (limit as f64) < input_prop.cardinality => limit as f64,
```
Comment (xudong963) [inline src/query/service/src/sql/planner/plans/limit.rs:80]: Not understand here.

File: src/query/service/src/sql/optimizer/rule/transform/rule_commute_join.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+use crate::sql::plans::LogicalInnerJoin;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::RelOp;
+
+/// Rule to apply commutivity of join operator.
+/// Since we will always use the right child as build side, this
+/// rule will help us measure which child is the better one.
+///
+/// TODO(leiysky): currently, we only support commutate for inner/cross join.
+/// Other join types will be added as soon as we implement them in Processor.
```
[the remark this answers is not in the document]
Comment (leiysky) [reply src/query/service/src/sql/optimizer/rule/transform/rule_commute_join.rs:31]: > A little confusion, other join types don't be implemented in processor?

For example, left(physically) semi join, right semi join, left anti join, right anti join.

[the remark this answers is not in the document]
Comment (xudong963) [reply src/query/service/src/sql/optimizer/rule/transform/rule_commute_join.rs:31]: got it

File: src/query/service/src/sql/planner/plans/limit.rs (hunk 2 above)
Comment (leiysky) [reply src/query/service/src/sql/planner/plans/limit.rs:80]: For example,
```sql
select * from numbers(10) limit 1; -- cardinality 1.0
select * from numbers(10) limit 20; -- cardinality 10.0
```

File: src/query/service/src/sql/optimizer/cascades/mod.rs [hunk 4]
```diff
@@ -15,115 +15,119 @@
...
             memo: Memo::create(),
             explore_rules: get_explore_rule_set(),
             implement_rules: get_implement_rule_set(),
+            cost_model: Box::new(DefaultCostModel),
+            best_cost_map: HashMap::new(),
         }
     }
 
-    #[allow(dead_code)]
     fn init(&mut self, expression: SExpr) -> Result<()> {
```
Comment (xudong963) [inline src/query/service/src/sql/optimizer/cascades/mod.rs:63]: I noticed you changed `expression` to `s_expr` in other places, maybe you'll be pleased to change it here

File: src/query/service/src/sql/optimizer/cascades/mod.rs [hunk 5]
```diff
@@ -15,115 +15,119 @@
...
         Ok(())
     }
 
-    #[allow(dead_code)]
-    pub fn optimize(&mut self, expression: SExpr) -> Result<SExpr> {
-        self.init(expression)?;
+    pub fn optimize(mut self, s_expr: SExpr) -> Result<SExpr> {
+        self.init(s_expr)?;
+
+        let root_index = self.memo.root().unwrap().group_index;
```
Comment (xudong963) [inline src/query/service/src/sql/optimizer/cascades/mod.rs:72]: Is `root` possibly None?

Comment (leiysky) [reply src/query/service/src/sql/optimizer/cascades/mod.rs:72]: It's impossible to be `None` after initialization, I will remove the `unwrap`.

File: src/query/service/src/sql/optimizer/cascades/mod.rs (hunk 4 above)
Comment (leiysky) [reply src/query/service/src/sql/optimizer/cascades/mod.rs:63]: Fixed.

**Review by xudong963 — APPROVED** [review]

File: src/query/catalog/src/table.rs [hunk 6]
```diff
@@ -224,7 +224,7 @@ pub enum NavigationPoint {
     TimePoint(DateTime<Utc>),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (Xuanwo) [inline src/query/catalog/src/table.rs:227]: This struct is safe to `Copy`.

File: src/query/datavalues/src/data_value.rs [hunk 7]
```diff
@@ -367,6 +368,23 @@ impl Ord for DataValue {
...
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
```
Comment (Xuanwo) [inline src/query/datavalues/src/data_value.rs:373]: This hash seems incorrect to me. Do we need to add `discriminant` for every enum variant?

Comment (leiysky) [reply src/query/datavalues/src/data_value.rs:373]: Yes, you are right.

File: src/query/catalog/src/table.rs (hunk 6 above)
Comment (leiysky) [reply src/query/catalog/src/table.rs]: Fixed

**Review by andylokandy — APPROVED** [review]

**Comment by leiysky** [comment]
@mergify update

**Pull request merged**